src/Controller/Component/FormProtectionComponent.php

```diff
@@ -0,0 +1,162 @@
+     * @param \Cake\Event\EventInterface $event An Event instance
+     * @return \Cake\Http\Response|null
+     */
+    public function startup(EventInterface $event): ?Response
+    {
+        $request = $this->getController()->getRequest();
+        $data = $request->getParsedBody();
+        $hasData = ($data || $request->is(['put', 'post', 'delete', 'patch']));
+
+        if (
+            !in_array($request->getParam('action'), $this->_config['unlockedActions'], true)
+            && $hasData
+            && $this->_config['validate']
+        ) {
+            $formProtector = new FormProtector();
+            $request->getSession()->start();
+            $isValid = $formProtector->validate(
+                $data,
+                $request->getRequestTarget(),
+                $request->getSession()->id()
+            );
+
+            if (!$isValid) {
+                return $this->validationFailure($formProtector);
+            }
+        }
+
+        $token = [
+            'unlockedFields' => $this->_config['unlockedFields'],
+        ];
+        $request = $request->withAttribute('formTokenData', [
+            'unlockedFields' => $token['unlockedFields'],
+        ]);
+
+        if (is_array($data)) {
+            unset($data['_Token']);
+            $request = $request->withParsedBody($data);
+        }
+
+        $this->getController()->setRequest($request);
+
+        return null;
+    }
+
+    /**
+     * Events supported by this component.
+     *
+     * @return array
+     */
+    public function implementedEvents(): array
+    {
+        return [
+            'Controller.startup' => 'startup',
+        ];
+    }
+
+    /**
+     * Throws a 400 - Bad request exception or calls custom callback.
+     *
+     * If `validationFailureCallback` config is specified, it will use this
+     * callback by executing the method passing the argument as exception.
+     *
+     * @param \Cake\Form\FormProtector $formProtector Form Protector instance.
+     * @return \Cake\Http\Response|null If specified, validationFailureCallback's response, or no return otherwise.
+     * @throws \Cake\Http\Exception\BadRequestException
+     */
+    protected function validationFailure(FormProtector $formProtector): ?Response
+    {
+        if ($this->_config['validationFailureCallback']) {
+            return $this->executeCallback($this->_config['validationFailureCallback'], [$formProtector]);
+        }
+
+        if (!Configure::read('debug')) {
+            throw new BadRequestException(static::DEFAULT_EXCEPTION_MESSAGE);
+        }
+
+        $errorMessage = $formProtector->getError() ?: static::DEFAULT_EXCEPTION_MESSAGE;
+        throw new BadRequestException($errorMessage);
+    }
+
+    /**
+     * Execute callback.
+     *
+     * @param callable $callback A valid callable
+     * @param array $params Params
+     * @return \Cake\Http\Response|null
+     */
+    protected function executeCallback(callable $callback, array $params): ?Response
```

Comment:
Should we require a closure? Users will have access to `Closure::fromCallable()`